### PR TITLE
UAR-745: Review trusts CYA

### PR DESCRIPTION
**Pull request by joshuahigginsonCH**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-745

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/controllers/update/update.manage.trusts.orchestrator.controller.ts [hunk 1]
```diff
@@ -60,6 +69,60 @@ const shouldGoToReviewTheTrust = (trustInReview: Trust) => {
   return !trustInReview.review_status?.reviewed_trust_details;
 };
 
+export const trustChangeHandler = async (req: Request, res: Response, next: NextFunction) => {
```
Comment (AndreKramer7654) [inline src/controllers/update/update.manage.trusts.orchestrator.controller.ts:72]: these changes are unrelated  to CYA? Do you just need to merge with main?

File: src/middleware/navigation/has.trust.middleware.ts [hunk 2]
```diff
@@ -28,6 +29,54 @@ const hasTrustWithId = (req: Request, res: Response, next: NextFunction, url: st
...
+        case TrusteeType.HISTORICAL:
+          hasTrusteeForChange = HISTORICAL.some(t => t.id === trusteeId);
+          break;
+        case TrusteeType.INDIVIDUAL:
+          hasTrusteeForChange = INDIVIDUAL.some(t => t.id === trusteeId);
+          break;
+        case TrusteeType.LEGAL_ENTITY:
+          hasTrusteeForChange = CORPORATE.some(t => t.id === trusteeId);
+          break;
+        default:
```
Comment (AndreKramer7654) [inline src/middleware/navigation/has.trust.middleware.ts:65]: no need for empty default case

File: src/middleware/navigation/has.trust.middleware.ts [hunk 3]
```diff
@@ -28,6 +29,54 @@ const hasTrustWithId = (req: Request, res: Response, next: NextFunction, url: st
...
+
+    if (!isTrustPresent) {
+      logger.infoRequest(req, NavigationErrorMessage);
+
+      return res.redirect(url);
+    }
+
+    let hasTrusteeForChange = false;
+
+    const HISTORICAL = appData[TrustKey]?.find(t => t.trust_id === trustId)?.HISTORICAL_BO ?? [];
```
Comment (AndreKramer7654) [inline src/middleware/navigation/has.trust.middleware.ts:51]: move these into case where they are needed/used

File: src/middleware/navigation/has.trust.middleware.ts [hunk 4]
```diff
@@ -28,6 +29,54 @@ const hasTrustWithId = (req: Request, res: Response, next: NextFunction, url: st
...
+          hasTrusteeForChange = INDIVIDUAL.some(t => t.id === trusteeId);
+          break;
+        case TrusteeType.LEGAL_ENTITY:
+          hasTrusteeForChange = CORPORATE.some(t => t.id === trusteeId);
+          break;
+        default:
+          break;
+    }
+
+    if (!hasTrusteeForChange && trusteeId) {
```
Comment (AndreKramer7654) [inline src/middleware/navigation/has.trust.middleware.ts:69]: seems a bit late to worry about trusteeId being present? Maybe handle that earlier

Comment (joshuahigginsonCH) [reply src/middleware/navigation/has.trust.middleware.ts:69]: Fixed

File: src/middleware/navigation/has.trust.middleware.ts (hunk 3 above)
Comment (joshuahigginsonCH) [reply src/middleware/navigation/has.trust.middleware.ts:51]: fixed

File: src/middleware/navigation/has.trust.middleware.ts (hunk 2 above)
Comment (joshuahigginsonCH) [reply src/middleware/navigation/has.trust.middleware.ts:65]: fixed

File: src/controllers/update/update.manage.trusts.orchestrator.controller.ts (hunk 1 above)
Comment (joshuahigginsonCH) [reply src/controllers/update/update.manage.trusts.orchestrator.controller.ts:72]: no these changes were introduced by me to handle the change links me and Cory just have some overlap

File: src/middleware/navigation/has.trust.middleware.ts (hunk 4 above)
Comment (joshuahigginsonCH) [reply src/middleware/navigation/has.trust.middleware.ts:69]: It was to handle the trust change link too but moved to a different approach

File: src/middleware/navigation/has.trust.middleware.ts [hunk 5]
```diff
@@ -28,6 +29,49 @@ const hasTrustWithId = (req: Request, res: Response, next: NextFunction, url: st
...
+      switch (trusteeType as TrusteeType){
+          case TrusteeType.HISTORICAL:
+            hasTrusteeForChange = (appData[TrustKey]?.find(t => t.trust_id === trustId)?.HISTORICAL_BO ?? []).some(t => t.id === trusteeId);
+            break;
+          case TrusteeType.INDIVIDUAL:
+            hasTrusteeForChange = (appData[TrustKey]?.find(t => t.trust_id === trustId)?.INDIVIDUALS ?? []).some(t => t.id === trusteeId);
+            break;
+          case TrusteeType.LEGAL_ENTITY:
+            hasTrusteeForChange = (appData[TrustKey]?.find(t => t.trust_id === trustId)?.CORPORATES ?? []).some(t => t.id === trusteeId);
+            break;
```
Comment (AndreKramer7654) [inline src/middleware/navigation/has.trust.middleware.ts:60]: would be good to have a default case that raises an error. 
default:  throw createAndLogErrorRequest(req, "Invalid trustee type");

File: src/utils/update/review_trusts.ts [hunk 6]
```diff
@@ -120,3 +120,19 @@ export const moveTrustOutOfReview = (appData: ApplicationData) => {
 
   appData.trusts?.push(trust);
 };
+
+export const putTrustInChangeScenario = (appData: ApplicationData, trustId: string, trusteeType?: string) => {
+  const trustForChangeScenario = appData.trusts?.splice(appData.trusts.findIndex(trust => trust.trust_id === trustId), 1)[0] as Trust;
```
Comment (AndreKramer7654) [inline src/utils/update/review_trusts.ts:125]: some of the values come from the user. So the trust_id could be not found. Or trustee type not something expected.

File: src/utils/update/review_trusts.ts [hunk 7]
```diff
@@ -120,3 +120,19 @@ export const moveTrustOutOfReview = (appData: ApplicationData) => {
...
+
+  appData.update?.review_trusts?.push(trustForChangeScenario);
+
+  const trustInChangeScenario = (appData.update?.review_trusts ?? [])[0];
```
Comment (AndreKramer7654) [inline src/utils/update/review_trusts.ts:129]: what would this be if [][0] is taken?

Comment (joshuahigginsonCH) [reply src/utils/update/review_trusts.ts:129]: fixed

File: src/utils/update/review_trusts.ts (hunk 6 above)
Comment (joshuahigginsonCH) [reply src/utils/update/review_trusts.ts:125]: The user will never be able to see or edit the URL as it is just an orchestrator call.

Comment (conuirchgov) [reply src/utils/update/review_trusts.ts:125]: We have middleware too on these routes, right? `hasTrustWithId` and `hasTrusteeWithId` or something similar, that'll verify the parameters given are valid before the request reaches the controller.

Comment (joshuahigginsonCH) [reply src/utils/update/review_trusts.ts:125]: Exactly @conuirchgov

File: src/routes/index.ts [hunk 8]
```diff
@@ -587,6 +587,24 @@ router.route(config.UPDATE_MANAGE_TRUSTS_ORCHESTRATOR_URL)
...
+    companyAuthentication,
+    navigation.hasTrustWithIdUpdate
+  )
+  .all(updateManageTrustsOrchestrator.trustChangeHandler);
+
+router.route(config.UPDATE_MANAGE_TRUSTS_ORCHESTRATOR_CHANGE_HANDLER_URL + config.TRUST_ID + '?' + config.TRUSTEE_TYPE + '?' + config.TRUSTEE_ID + '?')
+  .all(
+    isFeatureEnabled(config.FEATURE_FLAG_ENABLE_UPDATE_MANAGE_TRUSTS),
+    authentication,
+    companyAuthentication,
```
Comment (AndreKramer7654) [inline src/routes/index.ts:603]: please add hasTrustWithIdUpdate to this route

Comment (joshuahigginsonCH) [reply src/routes/index.ts:603]: The check for the trust is covered in the hasTrusteeWithId middleware below

Comment (AndreKramer7654) [reply src/routes/index.ts:603]: done withing hasTrusteeWithIdUpdate

**Review by conuirchgov — APPROVED** [review]

**Review by AndreKramer7654 — APPROVED** [review]

**Review by corybudd-ch — APPROVED** [review]
